go/sqltypes/result.go

```diff
@@ -185,13 +185,14 @@ func ResultsEqual(r1, r2 []Result) bool {
 // Every place this function is called, a comment is needed that explains
 // why it's justified.
 func MakeRowTrusted(fields []*querypb.Field, row *querypb.Row) []Value {
-	sqlRow := make([]Value, len(row.Lengths))
+	sqlRow := make([]Value, len(fields))
```

Comment:
For further safety, we should use the min of both lengths, and then iterate up to that point. It's quite possible that a row is shorter than the fields received.

Reply:
When I was talking about this with @setassociative, I suggested that in that case it would mean that you are removing a column. This is not backwards compatible, so we shouldn't worry about this.

However, after more thought, I think what you are suggesting makes sense. It could be possible that you change all your app to not use a column to be deprecated and then run into this bug because of a `select *`.

Reply:
It depends on which shard you hit first. Let's say you started adding a column. If you hit the shard that has the new column first, and then go to an old shard, the rows will be shorter. If you go the other way round, the fields will be shorter.

This actually gives rise to another problem: If the rows are shorter, then the client may crash. So, the more correct fix will be to extend the result with null values to match the field length.

Reply:
Oh yes. I did not think about the client crash. 

Adding a null value to match the fields, sounds good.

Reply:
I was trying to avoid a large behavioral change here (so not using min-length or NULL padding) because I'm very uncomfortable about increasing the number of scenarios where we return bad data (as opposed to just trading off where we fail).

Leaning only on the field length means that we will still panic in any case where we get caught in an intermediary state of removing a field but most of the times returning data in that state (I think?) would be distinctly bad anyway. As an example take the following table states:

Before DDL

| column name | type |
--|--
id|INT NOT NULL
name|VARCHAR NOT NULL
address|VARCHAR DEFAULT NULL
email | VARCHAR NOT NULL
age | INT DEFAULT NULL

After DDL

| column name | type |
--|--
id|INT NOT NULL
name|VARCHAR NOT NULL
email | VARCHAR NOT NULL
age | INT DEFAULT NULL

If we use the min(field, values) length to construct our result we will be passing back emails as addresses and NULL ages as (non-nullable) emails etc.

Padding with NULL values seems problematic for similar reasons.

In any case, I'm happy to swap over to min-length and/or NULL padding but wanted us to be explicitly okay with those failure modes.

Reply:
Most common use cases only add columns. So, adding null values as needed usually results in intended behavior.

But if we're more paranoid about correctness, we should go with @demmer's suggestion.

I personally prefer the heuristic approach because it's an improvement over what we have now, and it also solves the reported problem. But I won't be opposed to the more strict approach if people prefer that.

Reply:
It's true that most alters just add columns, but to richard's point they aren't always appended to the end, so if there's any kind of mismatch I still think it's better to be stricter and just return an error rather than papering over the problem with nulls.

Reply:
Sounds good.

Reply:
What I don't like about returning an error is that this is a pretty disruptive error that will be very hard to reason about for Vitess users. I wouldn't expect that adding a column could break existent queries. 

In our case, we can protect from this by banning `SELECT *` constructs in our codebase, but I don't think that would be a good expectation for other Vitess users. I think I would lean towards flexibility. 

@setassociative  had another idea the he proposed in: https://github.com/vitessio/vitess/issues/4661#issuecomment-466795191

I think that would be the best of both worlds. If we detect a mismatch in the fields, we error that query but immediately force a reload of the schema.

Reply:
It seems to me there are two related issues:

One is the tabletserver's cache of the field info not matching what it got back from mysql. @setassociative's idea of failing the query and triggering a reload seems like a good pragmatic idea for that.

The second is what happens when vtgate is processing a scatter query from shards with different field signatures. In this case it is possible that vttablet agrees with mysql, so that idea won't end up taking effect.

In both cases I think this is an unavoidable danger of using a distributed system and `select *` so we should strive towards correctness (failing the query) over "flexibility" (returning potentially garbage).

Reply:
I agree with @demmer. I would prefer to be correct and push people away from `SELECT *` than to guess.

Reply:
I vote for correctness as well.

Reply:
Sorry, I dropped the thread on this one -- I wasn't sure that the scatter aggregation code didn't already handle this case so I wanted to verify before filing that as a separate issue. It sounds like it doesn't though so I went ahead and filed https://github.com/vitessio/vitess/issues/4669.

Reply:
Should we consider ban `SELECT *` altogether?

Reply:
I strongly disagree that we should ban it as it would definitely discourage adoption.

I think being up front about the fact that during schema changes we might have situations where select * returns a transient error is acceptable, but banning it would be a step in the wrong direction.

Vitess or no vitess, schema changes and `select *` make for an unpredictable situation, but I wouldn't want to discourage adoption by making people rewrite their whole app.

Reply:
I was talking to @setassociative IRL. I think that with the expectation that the system will heal itself and eventually converge (which seems like it's what will happen if we do the pragmatic change + returning error while you are in this state), there is no need to be too strict with `SELECT *`.